src/model.rs

```diff
@@ -724,6 +724,138 @@ impl Collections {
+            .vehicle_journeys
+            .iter()
+            // Filtering out vehicle journeys without any stop time allowing us
+            // below to call stop_times[0] and stop_times[stop_times.len() - 1]
+            // safely.
+            .filter(|(_, vj)| !vj.stop_times.is_empty())
+            .map(|(idx, _)| idx)
+            .collect();
+        // Init all vehicle journeys with:
+        // - no drop-off on first stop
+        // - no pick-up on last stop
+        for vj_idx in &vj_idxs {
+            let mut vehicle_journey = self.vehicle_journeys.index_mut(*vj_idx);
+            let first_stop = &mut vehicle_journey.stop_times[0];
+            first_stop.drop_off_type = 1;
+            let stop_times_len = vehicle_journey.stop_times.len();
+            let last_stop = &mut vehicle_journey.stop_times[stop_times_len - 1];
+            last_stop.pickup_type = 1;
+        }
+        // Keep only vehicle journeys with a `block_id` (concerned with
+        // prolongation trip).
+        vj_idxs.retain(|vj_idx| self.vehicle_journeys[*vj_idx].block_id.is_some());
+        // Sort the vehicule journeys by block_id, and if equal, order them by
+        // departure time.
+        vj_idxs.sort_by(|vj1_idx, vj2_idx| {
+            let vj1 = &self.vehicle_journeys[*vj1_idx];
+            let vj2 = &self.vehicle_journeys[*vj2_idx];
+            match vj1.block_id.cmp(&vj2.block_id) {
+                Ordering::Equal => vj1.stop_times[0]
+                    .departure_time
+                    .cmp(&vj2.stop_times[0].departure_time),
+                ordering => ordering,
+            }
+        });
+        for (prev_vj_idx, next_vj_idx) in vj_idxs.iter().zip(vj_idxs.iter().skip(1)) {
+            let prev_vj = &self.vehicle_journeys[*prev_vj_idx];
+            let next_vj = &self.vehicle_journeys[*next_vj_idx];
+            if prev_vj.block_id != next_vj.block_id {
+                // we can discard two vehicle journeys which are not part of the
+                // same prolongation trip.
+                continue;
+            }
+            let stop_times_len = prev_vj.stop_times.len();
+            let last_stop = &prev_vj.stop_times[stop_times_len - 1];
+            let first_stop = &next_vj.stop_times[0];
+            if last_stop.stop_point_idx == first_stop.stop_point_idx {
+                // We can discard prolongation trip when the stop point is
+                // identical (see Example 1 above).
+                continue;
+            }
+            if last_stop.departure_time > first_stop.arrival_time {
+                // We can discard when timing overlaps between arrival of first
+                // vehicle journey and departure of next vehicle journey of a
+                // prolongation trip (see Example 2 above).
+                continue;
+            }
+            // We're now in Example 3 (see above), let's allow pick-up and
+            // drop-off on prolongation trip.
```

Comment:
Maybe add a comment on the fact it might be a `2` (ODT) originally here but it is lost.

Reply:
Done in 9d01051ae1b943af6b669f445179534b8e0d3ba6.